test/infrastructure/docker/docker/machine.go

```diff
@@ -159,13 +160,27 @@ func (m *Machine) ExecBootstrap(data string) error {
 		return errors.Wrap(err, "failed to decode machine's bootstrap data")
 	}
 
-	m.log.Info("Running machine bootstrap scripts")
-	lines, err := cloudinit.Run(cloudConfig, m.container.Commander)
+	commands, err := cloudinit.Commands(cloudConfig)
 	if err != nil {
-		m.log.Error(err, strings.Join(lines, "\n"))
+		m.log.Info("cloud config failed to parse", "cloud-config", string(cloudConfig))
 		return errors.Wrap(err, "failed to join a control plane node with kubeadm")
 	}
 
+	m.log.Info("Running machine bootstrap scripts")
+	var out bytes.Buffer
+	for _, command := range commands {
+		cmd := m.container.Commander.Command(command.Cmd, command.Args...)
+		cmd.SetStderr(&out)
+		if command.Stdin != "" {
+			cmd.SetStdin(strings.NewReader(command.Stdin))
+		}
+		err := cmd.Run()
+		if err != nil {
+			m.log.Info("Failed running command", "command", command, "stderr", out.String())
```

Comment:
might log command args as well?

Reply:
second thought. does it makes sense to think about how to trace/log the cloud-init settings that generated the command?
without this, it might be hard to trace back the source from a failing error...

Reply:
`command` will log the command, args and stdin if applicable.

I don't think logging the cloud-init is necessary. Since the cloud-init is really opaque I have not found that level of debugging useful. The interesting part is figuring out which command failed and why, not what generated the command.

Reply:
kk, fair point